### PR TITLE
Add universal builds on macOS

### PR DESCRIPTION
**Pull request by juliangruber**
This adds arm64 support.

Universal artefacts should be better than separate arm64 and x64 builds:
- no need for the user to choose
- release asset urls don't change, breaking auto update

Closes https://github.com/filecoin-station/core/pull/150

### CODE REVIEW

**Comment by juliangruber** [comment]
`npm run package` failed locally:

```
fatal error: /Library/Developer/CommandLineTools/usr/bin/lipo: /private/var/folders/ph/cz_m_sjn3ys1cnnsmx5058tw0000gn/T/electron-universal-cSLgEL/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau and /Users/julian/dev/filecoin-station/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau/bacalhau have the same architectures (arm64) and can't be in the same fat output file
```

**Comment by juliangruber** [comment]
Different errors with separate targets:

> Cannot cleanup: 
>
> Error #1 --------------------------------------------------------------------------------
Error: hdiutil process failed ERR_ELECTRON_BUILDER_CANNOT_EXECUTE
Exit code:
1
    at ChildProcess.<anonymous> (/Users/julian/dev/filecoin-station/desktop/node_modules/builder-util/src/util.ts:250:14)
    at Object.onceWrapper (node:events:626:26)
    at ChildProcess.emit (node:events:511:28)
    at maybeClose (node:internal/child_process:1098:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5)
>
> Error #2 --------------------------------------------------------------------------------
Error: hdiutil process failed ERR_ELECTRON_BUILDER_CANNOT_EXECUTE
Exit code:
1
    at ChildProcess.<anonymous> (/Users/julian/dev/filecoin-station/desktop/node_modules/builder-util/src/util.ts:250:14)
    at Object.onceWrapper (node:events:626:26)
    at ChildProcess.emit (node:events:511:28)
    at maybeClose (node:internal/child_process:1098:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5)  failedTask=build

**Comment by juliangruber** [comment]
Running electron builder twice, once per arch, works locally

**Comment by juliangruber** [comment]
Ok, we have to change the release filename convention, breaking updates (I assume).

**Comment by juliangruber** [comment]
@bajtos this will break auto updates. But this will also add proper arm64 support 🙏

**Comment by bajtos** [comment]
@juliangruber could you please check that the arm64 artefacts produced by GitHub Action CI runs (executed on x86) contain zinniad & bacalhau binaries for arm64? I strongly suspect they will contain x86 versions.

> fatal error: /Library/Developer/CommandLineTools/usr/bin/lipo: /private/var/folders/ph/cz_m_sjn3ys1cnnsmx5058tw0000gn/T/electron-universal-cSLgEL/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau and /Users/julian/dev/filecoin-station/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau/bacalhau have the same architectures (arm64) and can't be in the same fat output file

Assuming this error is coming from an arm64 machine like Apple M1/M2, I think the error message tells us we are using the same arm64 flavour of bacalhau binary for both x86 and arm64 flavours of Station Desktop.

**Comment by bajtos** [comment]
You can download dist packages from GHA Summary page for our CI workflow, see e.g. https://github.com/filecoin-station/desktop/actions/runs/5242546465

**Comment by bajtos** [comment]
Not necessarily. I think electron-autoupdater is looking at `latest-mac.yml` to find out what to do. AFAICT from the CI "package" step output, we are still producing only one YML file for Mac.

**Comment by bajtos** [comment]
I downloaded [dist-macos-latest](https://github.com/filecoin-station/desktop/suites/13535047297/artifacts/744398108), unpacked "Filecoin Station-arm64.zip" and found this:

```
❯ file MacOS/Filecoin\ Station
MacOS/Filecoin Station: 
  Mach-O 64-bit executable arm64
❯ file "Contents/Resources/app.asar.unpacked/node_modules/@filecoin-station/core/modules/zinnia/zinniad"
Contents/Resources/app.asar.unpacked/node_modules/@filecoin-station/core/modules/zinnia/zinniad: 
  Mach-O 64-bit executable x86_64
```

**Comment by juliangruber** [comment]
> @juliangruber could you please check that the arm64 artefacts produced by GitHub Action CI runs (executed on x86) contain zinniad & bacalhau binaries for arm64? I strongly suspect they will contain x86 versions.
> 
> > fatal error: /Library/Developer/CommandLineTools/usr/bin/lipo: /private/var/folders/ph/cz_m_sjn3ys1cnnsmx5058tw0000gn/T/electron-universal-cSLgEL/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau and /Users/julian/dev/filecoin-station/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau/bacalhau have the same architectures (arm64) and can't be in the same fat output file
> 
> Assuming this error is coming from an arm64 machine like Apple M1/M2, I think the error message tells us we are using the same arm64 flavour of bacalhau binary for both x86 and arm64 flavours of Station Desktop.

This error came from both x64 and arm64 machines

**Comment by juliangruber** [comment]
> Interesting, this is after `npm run package`:
> 
> ```
> $ file dist/mac-arm64/Filecoin\ Station.app/Contents/Resources/app.asar.unpacked/node_modules/@filecoin-station/core/modules/zinnia/zinniad 
> (...): Mach-O 64-bit executable arm64
> ```
> 
> But also:
> 
> ```
> $ file dist/mac/Filecoin\ Station.app/Contents/Resources/app.asar.unpacked/node_modules/@filecoin-station/core/modules/zinnia/zinniad 
> (...): Mach-O 64-bit executable arm64
> ```

Turns out currently we're always bundling module executables with the same architecture as the builder host. Electron builder's rebuilding of native dependencies doesn't target Core (yet)

**Comment by bajtos** [comment]
NOICE! 🎊 

Could you please check why `ubuntu-latest` build is failing?

**Comment by juliangruber** [comment]
> Could you please check why ubuntu-latest build is failing?

External API error 🤷‍♂️  Unrelated

**Comment by bajtos** [comment]
I would like to download the macOS universal installer created by our GHA workflow to verify that all works correctly. To do that, I need a successful workflow run. I restarted the failed job.

**Comment by bajtos** [comment]
Nice! When building an universal binary, electron-builder is smart enough to run Station Core's post-install script in different locations:

```
Rebuild Station Core for arch=x64
(...)
[spark]  ✓ /Users/bajtos/src/filstation/desktop/dist/mac-universal--x64/Filecoin Station.app/Contents/Resources/core/modules/spark
[bacalhau] ✓ /Users/bajtos/src/filstation/desktop/dist/mac-universal--x64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau
[zinnia] ✓ /Users/bajtos/src/filstation/desktop/dist/mac-universal--x64/Filecoin Station.app/Contents/Resources/core/modules/zinnia
(...)
Rebuild Station Core for arch=arm64
(...)
[spark]  ✓ /Users/bajtos/src/filstation/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/spark
[zinnia] ✓ /Users/bajtos/src/filstation/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/zinnia
[bacalhau] ✓ /Users/bajtos/src/filstation/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau
```

The local files used by `npm start`, e.g. `core/modules/zinnia/zinniad`, are not touched 👍🏻 

(I was concerned that re-running the post-install script inside `core/modules` would leave the local repo with binaries for a wrong architecture. That's not the case.)

**Comment by bajtos** [comment]
@juliangruber the packaging step failed on macOS, see https://github.com/filecoin-station/desktop/actions/runs/5268127373/jobs/9524799363#step:9:454

```
fatal error: /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/electron-universal-X16CUn/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau and /Users/runner/work/desktop/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau/bacalhau have the same architectures (arm64) and can't be in the same fat output file  failedTask=build stackTrace=Error: Command failed with a non-zero return code (1):
lipo /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/electron-universal-X16CUn/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau /Users/runner/work/desktop/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau/bacalhau -create -output /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/electron-universal-X16CUn/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            fatal error: /Applications/Xcode_14.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/electron-universal-X16CUn/Tmp.app/Contents/Resources/core/modules/bacalhau/bacalhau and /Users/runner/work/desktop/desktop/dist/mac-universal--arm64/Filecoin Station.app/Contents/Resources/core/modules/bacalhau/bacalhau have the same architectures (arm64) and can't be in the same fat output file
    at ChildProcess.<anonymous> (/Users/runner/work/desktop/desktop/node_modules/@malept/cross-spawn-promise/src/index.ts:172:16)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at Socket.<anonymous> (node:internal/child_process:449:11)
    at Socket.emit (node:events:513:28)
    at Pipe.<anonymous> (node:net:322:12)
```

FWIW, `npm run package` works well on my machine 🤔

**Comment by juliangruber** [comment]
> I would like to download the macOS universal installer created by our GHA workflow to verify that all works correctly. To do that, I need a successful workflow run. I restarted the failed job.

I don't know if they will work for you, because of code signing. The result of `npm run package` on your machine should work.

**Comment by bajtos** [comment]
> > I would like to download the macOS universal installer created by our GHA workflow to verify that all works correctly. To do that, I need a successful workflow run. I restarted the failed job.
> 
> I don't know if they will work for you, because of code signing. The result of `npm run package` on your machine should work.

Is it because we are not notarising build artifacts created by pull requests? If that's the case, then I think we should publish a pre-release version to verify that the new universal installer is correctly notarised.

**Comment by juliangruber** [comment]
My assumption was wrong, bacalhau does have arm64 builds, something in the build logic is off

**Comment by juliangruber** [comment]
https://github.com/filecoin-station/desktop/releases/tag/v0.20.0-0

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -176,6 +176,7 @@ jobs:
           APPLEID: ${{ secrets.apple_id }}
           APPLEIDPASS: ${{ secrets.apple_id_pass }}
           APPLETEAMID: ${{ secrets.apple_team_id }}
+          USE_HARD_LINKS: false
```
Comment (bajtos) [inline .github/workflows/ci.yml:179]: 👌🏻

**Comment by bajtos** [comment]
Testing https://github.com/filecoin-station/desktop/releases/tag/v0.20.0-0 and all looks great so far!

After I installed the app, macOS asked me if I wanted to run an app downloaded from the internet. I confirmed and the Station started.

The job counter is steadily going up.

```
❯ file /Applications/Filecoin\ Station.app/Contents/MacOS/Filecoin\ Station
/Applications/Filecoin Station.app/Contents/MacOS/Filecoin Station: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64:Mach-O 64-bit executable arm64]
/Applications/Filecoin Station.app/Contents/MacOS/Filecoin Station (for architecture x86_64):	Mach-O 64-bit executable x86_64
/Applications/Filecoin Station.app/Contents/MacOS/Filecoin Station (for architecture arm64):	Mach-O 64-bit executable arm64
```

Station and Zinnia processes run the ARM (Apple) flavour (see the last column on my screenshots).

Screenshot 2023-06-19 at 10 06 18
Screenshot 2023-06-19 at 10 06 29

@juliangruber I think we are good to go and publish a new stable release 👍🏻

**Comment by juliangruber** [comment]
Fantastic, thank you for validating!

**Comment by juliangruber** [comment]
https://github.com/filecoin-station/desktop/releases/tag/v0.20.0

**Pull request merged**